---
title: "ITS primers and databeses 2"
output: html_notebook
---
### Primers table
Collapsed table.  
Table includes original sequences with ambiguous letters and variant numbers.  
```{r collapsed primers}
primers <- data.frame(Names = rep(names(primers_names_len), primers_names_len), Seq = unlist(primers_exp))
primers[grep("ITS4|NLC|ITS2", primers$Names), "direction"] <- "Reverse"
primers[grep("ITS4|NLC|ITS2", primers$Names, invert = T), "direction"] <- "Forward"
primers[grep("ITS7|ITS9|ITS3|ITS2|ITS86|Rust", primers$Names), "region"] <- "5.8S"
primers[grep("ITS4|NLC|LR|NL", primers$Names), "region"] <- "LSU_28S"
primers[grep("ITS2|58A", primers$Names), "region"] <- "SSU_18S"

primers_coll <- primers %>%
  mutate(Sequence = primers_sequences[match(primers$Names,names(primers_sequences))]) %>%
  dplyr::group_by(Names, region, direction, Sequence) %>%
  dplyr::summarise(No.variant = length(Seq))
primers_coll
```

### FINAL TABLE

Here we combine all our data and make final report table.

First to estimate percentages in each group we need to get values of total numbers.
```{r totals}

tatal_list <- function(tax) {
  list(
total_len = length(unique(tax$samples)),
total_orders = length(unique(tax$order)),
total_spp = length(unique(tax$species)),
total_genus = length(unique(tax$genus)),
total_fam = length(unique(tax$family)),
total_rust = sum(!is.na(match(tax$order,"Pucciniales"))),
total_asco = sum(!is.na(match(tax$phylum,"Ascomycota"))),
total_basidio = sum(!is.na(match(tax$phylum,"Basidiomycota")))
)
}
```
Summary table function.
```{r table function}
sum_table <- function(fshtax, totals) {
  fshtax.primers <- fshtax %>%
  mutate(No.variant = primers_names_len[match(primers,names(primers_names_len))],
         Rust = match(order,"Pucciniales"),
         Asco = match(phylum,"Ascomycota"),
         Basidio = match(phylum,"Basidiomycota")) %>%
  dplyr::group_by(primers) %>%
  dplyr::summarise(
    No.variant = unique(No.variant),
    No.samples = length(unique(samples)),
    efficiency = paste(round(sum(No.samples)/unique(No.variant)), " (", round(((sum(No.samples)/unique(No.variant)/totals[[1]])*100),2), "%)", sep = ""),
    Rust = paste(sum(!is.na(Rust)), " (", round((sum(!is.na(Rust))/totals[[6]])*100 ,2), "%)",  sep = ""),
    Asco = paste(sum(!is.na(Asco)), " (", round((sum(!is.na(Asco))/totals[[7]])*100 ,2), "%)",  sep = ""),
    Basidio = paste(sum(!is.na(Basidio)), " (", round((sum(!is.na(Basidio))/totals[[8]])*100 ,2), "%)",  sep = ""),
    No.orders = length(unique(order)),
    No.fam = length(unique(family)),
    No.genus = length(unique(genus)),
    No.spp = length(unique(species)),
    No.samples = paste(sum(No.samples), " (", round((sum(No.samples)/totals[[1]])*100, 2), "%)", sep = ""), 
    No.orders = paste(sum(No.orders), " (", round((sum(No.orders)/totals[[2]])*100, 2), "%)", sep = ""),
    No.fam = paste(sum(No.fam), " (", round((sum(No.fam)/totals[[5]])*100, 2), "%)", sep = ""),
    No.genus = paste(sum(No.genus), " (", round((sum(No.genus)/totals[[4]])*100, 2), "%)", sep = ""),
    No.spp = paste(sum(No.spp), " (", round((sum(No.spp)/totals[[3]])*100, 2), "%)", sep = "")
    ) %>%
  mutate(Sequence = primers_sequences[match(primers,names(primers_sequences))])

write.table(fshtax.primers, file = "fished_primers_rep.tab", append = F, row.names = F, sep = "\t", quote = F)
fshtax.primers
}
```

Now we create joined taxonomy and fished data tables.
```{r joined tables, warning=FALSE, message=FALSE}
fshtax_28S <- inner_join(fsh_28S, taxgb, by = "samples")
fshtax_UNITE <- inner_join(fsh_UNITE, tax_unite, by = "samples")
```
Totals:  
```{r total values}
total_28S <- tatal_list(tax = taxgb)
total_UNITE <- tatal_list(tax = tax_unite)
```

Final summary table:
```{r final}
fished.primers.28S <- sum_table(fshtax = fshtax_28S, totals = total_28S)
fished.primers.unite <- sum_table(fshtax = fshtax_UNITE, totals = total_UNITE)
```

### BUBBLE PLOT
This function is similar to summary table.
```{r plotdata function}

bubble_data <- function(fshtax, totals) {
  bbtbl <- fshtax %>%
  mutate(No.variant = primers_names_len[match(primers,names(primers_names_len))],
         Rust = match(order,"Pucciniales"),
         Asco = match(phylum,"Ascomycota"),
         Basidio = match(phylum,"Basidiomycota")) %>%
  dplyr::group_by(primers) %>%
  dplyr::summarise(
    No.variant = unique(No.variant),
    Total.samples = length(unique(samples))/totals[[1]],
    total.efficiency = sum(length(unique(samples)))/unique(No.variant)/totals[[1]],
    Rust = sum(!is.na(Rust))/totals[[6]],
    Asco = sum(!is.na(Asco))/totals[[7]],
    Basidio = sum(!is.na(Basidio))/totals[[8]])
bbtbl
}
```
Melting table for ggplot2.
```{r melting data}
bubble.28S <- bubble_data(fshtax = fshtax_28S, totals = total_28S)
bubble.unite <- bubble_data(fshtax = fshtax_UNITE, totals = total_UNITE)

bubble.28S$DB <- "rRNA_28S"
bubble.unite$DB <- "UNITE"
bbdt <- rbind(bubble.28S, bubble.unite)
bbdt <- bubble.28S
bbdt[grep("ITS4|NLC|NLB", bbdt$primers), "direction"] <- "Reverse"
bbdt[grep("ITS4|NLC|NLB", bbdt$primers, invert = T), "direction"] <- "Forward"
bbdt <- gather(bbdt, variable, value, 3,5:7)
bbdt <- as.data.frame(bbdt)
bbdt <- bbdt[bbdt$value != "NaN",]
bbdt$variable <- factor(bbdt$variable, levels = sort(unique(bbdt$variable), decreasing = F))
bbdt

```
```{r colors and filters}
clr = brewer.pal(length(unique(bbdt$variable)), "Set1")
names(clr) <- levels(bbdt$variable)

# Uncomment if you need to filter data by x values
#bbdt <- bbdt[bbdt$value > 0.5,]
```


Plotting bubble data.   
```{r ggplot, fig.width=12, fig.height=6}

bbdt_plot <- ggplot(bbdt, aes(value, primers, size = value, fill = variable, shape = DB)) +
  geom_point(alpha = 0.6, colour = "#636363") + 
   labs(y = "", x = "Group efficiency") +
  scale_size_area(max_size=12, guide=guide_legend(title="Per primer efficiency", keyheight=0)) +
  scale_fill_manual(values = clr, name = "Taxons") +
  scale_shape_manual(values = c(21,22,23), guide=guide_legend(override.aes=aes(size=8)), name = "DataBases") +
  guides(fill = guide_legend(override.aes = list(size = 12, colour = clr), reverse = F)) +
  #facet_grid(direction~.,  scales = "free", space = "free")
   theme(panel.background=element_rect(fill="white"),
        legend.key = element_rect(fill = "white", color = "white"),
        #plot.title = element_text(lineheight=.8, face="bold"),
        #title=element_text(size=15,colour="black"),
        strip.background = element_rect(colour = "white", fill = "white"),
        strip.text = element_text(size = 12, face = "bold", hjust = 0),
        #strip.background = element_rect(colour="black", fill="#ffffff"),
        #axis.title.y=element_text(size=18,colour="black", angle = 90),
        axis.title.x=element_text(size=14,colour="black"),
        axis.line = element_line(colour="black"),
        text=element_text(size=10,colour="black"),
        panel.grid.major=element_line(size=0.5,colour="gray80",linetype = "dotted"),
        panel.grid.minor=element_line(size=0.5,colour="gray80",linetype = "dotted"),
        panel.border = element_rect(fill = NA, colour = "#ffffff"),
        #legend.position=c(1.2,0.7),
        legend.position = "right",
        legend.text = element_text(size = 12),
        axis.text = element_text(size = 14, colour = "black", face = "italic"))
         #strip.background = element_rect(colour="#ffffff", fill="#f0f0f0"))
bbdt_plot
#saveRDS(bbdt_plot, file = "../presentations/bbdt_plot.rds")
```
Saving images  
```{r saving picture, results="hide"}
png(filename = "primers.png", height = 500, width = 1000, units = "px")
bbdt_plot
dev.off()
```

