---
title: "ITS primers and databeses 1"
output: html_notebook
---
## Data preporation: databeses, primers, matching, taxonomy.

### Introduction 
ITS region primers are not 100% effective. Due to that various important fungal species tend escape identification. Here we present an <i>in silico</i> analysis of how effective are various ITS primer. 
<h3>Databases</h3>
Databse files are located in 'db' directory.  
FASTA files:  

* 'UNITEv7_public_20.11.2016.fasta' -- file containing latest UNITE+ databse. There is no taxonomy file for this one, but it can be easily extracted from sequence names (see below).
* 'fungi.ITS.fna' -- GeneBank data for ITS.
* 'fungi.28S.fna' -- GeneBank data for 28S.
* 'isham.fasta' -- ISHAM database. Human fungal pathogens ITS database. http://its.mycologylab.org/
* 'UNITEv6_sh_99_s.fasta' -- older version of UNITE.
* 'Rust_trimmed_for_ITS4_primer_analysis.fasta' --  our in-house rust sequences for the ITS4 analysis only. They are fairly short because I trimmed off the ITS2 region. So, they start at the beginning of the 28S which includes the priming site for ITS4 and all the variants, and I assume also the sites for the other reverse primers being tested.
* 'GenBank_Pucciniales_28SwithITS .fasta' -- The other has all the GenBank sequences I downloaded using these two queries:  
(1300:4000[Sequence Length]) AND Pucciniales[orgn] AND ("internal transcribed spacer 2" NOT "internal transcribed spacer 1"[Title]) AND "28S ribosomal" NOT "whole genome" NOT uncultured NOT intron NOT "environmental" NOT "clone" NOT "shotgun" NOT "sp.” NOT 18S[title]
(800:6000[Sequence Length]) AND Pucciniales[orgn] AND ("5.8S"[Title]) AND "28S ribosomal" NOT "whole genome" NOT uncultured NOT intron NOT "environmental" NOT "clone" NOT "shotgun" NOT "sp.” NOT 18S[title]
* 'RS_cerealrusts_with5.8S.fasta' -- cereal rust sequences.

TAXONOMY files:

* 'UNITEv6_sh_99_s.tax' -- tax file for older UNITE.
* 'ISHAM.tax' -- tax file for ISHAM db, was provided slong with fasta file.
* GeneBank -- to extract GneBank taxonomy see function below.

```{r libraries,  results='hide'}
library(Biostrings)
library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(RColorBrewer)
library(taxize)
```

### Primers 

All primers are stored in 'ITS2_Primers.fasta'
Primers used in this study expanded using Biostrings.
```{r expantion}
expand.dna<-function(x){
  # IUPAC_CODE_MAP comes from Biostrings package.
  code <- as.list(IUPAC_CODE_MAP)
  code<- lapply(code, function(y) strsplit(y, "")[[1]] )
  y <-strsplit(as.character(x), "")[[1]]
  n <- match(y, names(code))
  z <- expand.grid(code[n] )
  apply(z, 1, paste, collapse="")
}
```

Expanded primers table with all annotations.  
```{r primers}
primers_or <- readDNAStringSet("../../../UNITE/ITS2_Primers.fasta")
primers_exp <- lapply(primers_or, expand.dna)
primers_names_len <- sapply(primers_exp, length)
primers_sequences <- as.vector(primers_or)
primers <- DNAStringSet(unlist(primers_exp))
names(primers) <- rep(names(primers_names_len), primers_names_len)
primers_rev <- reverseComplement(primers)
primers <- append(primers, primers_rev, after = length(primers))
# Uncomment following line if you want to write full set of primers.
# writeXStringSet(primers, filepath = "primers_expanded.fasta")
```
### Matching
Matching primers against database.
As 'oligofishing' excludes sequence from the pool, as long as it was matched, we need to run it in a loop matching each primer individually.  
Script may take few hours to run with big database (like UNITE), so it is advised to run this piece separately.  
This script is to run on Workstation.
```{r matching, results='hide'}
for(i in 1:length(primers)) {
  writeXStringSet(primers[i], filepath = "primer_expanded.fasta", append = F)
  
  #system("~/bean/oligofishing -h primer_expanded.fasta -p UNITE_public_20.11.2016.fasta >> UNITE_fished.fasta")
  system("~/bin/oligofishing/oligofishing -h primer_expanded.fasta -p ../../../UNITE/fungi.28SrRNA.fna >> ../../../UNITE/rRNA28S_fished.fasta")
}
  
  system("grep \">\" ../../../UNITE/rRNA28S_fished.fasta | sed 's/>//g' > ../../../UNITE/rRNA28S_fished.tab")
  file.remove("primer_expanded.fasta")
```
#### Fished data
Load fishing results for entire UNITE data base.  
```{r fished split function}
fished_split <- function(names_file) {
  fsh_seqs <- fread(names_file, sep = "\t", col.names = "samples")
  fsh <- strsplit(as.character(unique(as.character(fsh_seqs$samples))), '|', fixed = T)
fsh
}

```

Creating fished data frame.  
```{r fished2}
fsh <- fished_split("UNITEv7_public_fished.tab")
fsh_28S <- fished_split("../../../UNITE/rRNA28S_fished.tab")
fsh_UNITE <- data.frame(samples = sapply(fsh, "[", 1), primers = sapply(fsh, function(x) x[length(x)]))
fsh_28S <- data.frame(samples = sapply(fsh_28S, "[", 4), primers = sapply(fsh_28S, function(x) x[length(x)]))
```
### Taxonomy
Here contents of column 'samples' must corrspond to column 'samples' in matched output data.  

#### Two functions
1. Following function is to extract taxonomy from data comming from .tax files or similar format. In our example this function works for Case 1 and Case 2 taxonomy.  
This function creates separate columns for taxonomic groups.  
You need to provide:  
* taxa dataframe. This dataframe must contain following columns:
    + 'samples', cantaining names for matched sequences. 
    + 'taxonomy', containig list of vectors with taxonomy data for each sample.
```{r taxonomicGroups}
tax_extract <- function(taxa = tax) {
  taxon_groups <- c("^s__*", "^g__*", "^f__*", "^o__*", "^c__*", "^p__*", "^k__*")
  names(taxon_groups) <- c("species", "genus", "family", "order", "class", "phylum", "kingdom")

        for(i in 1:length(taxon_groups)) {
          taxon <- function(x) {
            y <- x[grep(taxon_groups[i], x)]
            if(length(y) == 0) {
              y = as.character(NA)
            } else {
              y
            }
          }
          taxa[names(taxon_groups)[i]] <- sapply(taxa$taxonomy, taxon)
        }
  tax
}
```

2. Function to extract taxonomy from GeneBank. Case 3.  
Here you need to provide vector with GB accession numbers. The same ids must be used in following steps. Inside the function there is vector containing rank names to be used for all data.  
For this function you need to provide (see Case 3 below):  
* Classification data, out put from 'taxsize::classification()'.
* Accession numbers vector used to retrieve initial classification.
```{r taxonomy function GB}
taxonomy_gb <- function(classification_data = funclsf, acc = acc_ids) {
          taxon_groups <- factor(c("species", "genus", "family", "order", "class", "phylum", "kingdom"))
          levels(taxon_groups) <- c("species", "genus", "family", "order", "class", "phylum", "kingdom")
          classif_fun <- function(x) {
            y = x[match(taxon_groups, x$rank),-3]
            z = data.frame(y, stringsAsFactors = F, row.names = NULL)
            z$rank <- taxon_groups
            z = tidyr::spread(z, key = rank, value = name)
            z
          }
      
      clsf <- lapply(classification_data, classif_fun)
      tax_gb <- bind_rows(clsf) 
      tax_gb$tax_id <- names(clsf)
      tax_gb <- mutate(tax_gb, samples = names(acc)[match(tax_gb$tax_id,acc)])
tax_gb     
}
```
#### Case 1
.tax file provided  
```{r tax file taxonomy}
tax <- fread("UNITEv6_sh_99_s.tax", sep = "\t", col.names = c("samples", "taxonomy"))
tax <- as.data.frame(tax)
tax$taxonomy <- strsplit(as.character(tax$taxonomy), ';', fixed = T)
tax <- tax_extract(taxa = tax)
```

#### Case 2
If there is no .tax file we can extract taxonomy from UNITE sequence names.
This allows to create the same data.frame as if there were .tax file.
```{r UNITE taxonomy}
tax <- fread("UNITEv7_public.tab", sep = "\t", col.names = "samples")
tax <- as.data.frame(tax)
tax$samples <- strsplit(as.character(unique(as.character(tax$samples))), '|', fixed = T)
tax$taxonomy <- sapply(tax$samples, "[", 2)
tax$taxonomy <- strsplit(as.character(tax$taxonomy), ';', fixed = T)
tax$samples <- sapply(tax$samples, "[", 1)
tax_unite <- tax_extract(taxa = tax)
```
#### Case 3.
When you need to extract taxomic data from GeneBank.  
First we retrieve taxonomy ids from NCBI using accession numbers from sequences names in fasta file. This is done with packages 'taxize' and 'data.table'.
```{r taxize, cache=TRUE}
# File with extracted sequences names.
its_refseq <- read.table(file = "../../../UNITE/fungi.28SrRNA.tab", sep = "|")
# naming column with accesession numbers.
colnames(its_refseq)[4] <- "acc"
ids = genbank2uid(id = its_refseq$acc)
funclsf <- classification(ids, db = "ncbi")
```
Now we create a named vector with taxon ids and names corresponding to accession numbers. In output data from matching column 'samples' must contain these ids.  
```{r samples}
acc_ids = as.vector(ids)
names(acc_ids) <- its_refseq$acc
```

Now we can use function to create data.frame with taxonomic information. 
```{r taxonomy GB}
taxgb <- taxonomy_gb(classification_data = funclsf, acc = acc_ids)
taxgb
```

